File: benchmarks/pandas/bench_timedelta_index_ops.py
"""Benchmark: TimedeltaIndex sort / unique / shift / min / max on 1k-element index."""
import json, time
import pandas as pd
import numpy as np

SIZE = 1_000
WARMUP = 5
ITERATIONS = 100

deltas = [pd.Timedelta(days=(i * 13) % 365, hours=i % 24) for i in range(SIZE)]
idx = pd.TimedeltaIndex(deltas)
shift_by = pd.Timedelta(days=1)
threshold = pd.Timedelta(days=100)

for _ in range(WARMUP):
    idx.sort_values()
    idx.unique()
    idx + shift_by
    idx[idx < threshold]
    idx.min()
    idx.max()

times = []
for _ in range(ITERATIONS):
    t0 = time.perf_counter()
    idx.sort_values()
    idx.unique()
    idx + shift_by
    idx[idx < threshold]
    idx.min()
    idx.max()
    times.append((time.perf_counter() - t0) * 1000)

total_ms = sum(times)
print(json.dumps({"function": "timedelta_index_ops", "mean_ms": round(total_ms / ITERATIONS, 3), "iterations": ITERATIONS, "total_ms": round(total_ms, 3)}))

File: benchmarks/tsb/bench_timedelta_index_ops.ts
/**
 * Benchmark: TimedeltaIndex.sort / unique / shift / filter / min / max — operations on 1k-element TimedeltaIndex.
 * Outputs JSON: {"function": "timedelta_index_ops", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Timedelta, TimedeltaIndex } from "../../src/index.ts";

const SIZE = 1_000;
const WARMUP = 5;
const ITERATIONS = 100;

// Shuffled, with some duplicates
const deltas = Array.from({ length: SIZE }, (_, i) =>
  Timedelta.fromComponents({ days: (i * 13) % 365, hours: i % 24 }),
);
const idx = TimedeltaIndex.fromTimedeltas(deltas);
const shiftBy = Timedelta.fromComponents({ days: 1 });
const threshold = Timedelta.fromComponents({ days: 100 });

for (let i = 0; i < WARMUP; i++) {
  idx.sort();
  idx.unique();
  idx.shift(shiftBy);
  idx.filter((td) => td.totalDays < threshold.totalDays);
  idx.min();
  idx.max();
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  idx.sort();
  idx.unique();
  idx.shift(shiftBy);
  idx.filter((td) => td.totalDays < threshold.totalDays);
  idx.min();
  idx.max();
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "timedelta_index_ops",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
